cleanup_utils/aws_cleanup.py

```diff
@@ -0,0 +1,227 @@
+        """
+        self.age_limit = age_limit
+        self.dry_run = dry_run
+        self.cleaned_up_hashes = []
+
+        self.session = boto3.Session(
+            aws_access_key_id=aws_access_key_id,
+            aws_secret_access_key=aws_secret_access_key,
+        )
+        self.iam_client = self.session.client('iam')
+        self.s3_client = self.session.client('s3')
+        self.s3_resource = self.session.resource('s3')
+
+    def delete_bucket(self, bucket_name):
+        """
+        Deletes a S3 bucket and all of it's files
+        """
+        if self.dry_run:
+            return
+        try:
+            logger.info("Deleting bucket %s.", bucket_name)
+            bucket = self.s3_resource.Bucket(bucket_name)
+            bucket.object_versions.delete()
+            bucket.objects.all().delete()
+            bucket.delete()
+        except self.s3_client.exceptions.NoSuchBucket:
+            # Ignore if the bucket doesn't exist
+            logger.warning("The bucket %s did not exist, skipping.", bucket_name)
+        except botocore.exceptions.ClientError as e:
+            # Ignore if the bucket doesn't exist
+            logger.error(
+                "Untreated error when deleting %s: %s",
+                bucket_name,
+                e.response['Error']['Message']
+            )
+
+    def delete_user_policy(self, username, policy_name):
+        """
+        Deletes policy associated to user
+        """
+        if self.dry_run:
+            return
+        self.iam_client.delete_user_policy(
+            UserName=username,
+            PolicyName=policy_name
+        )
+
+    def delete_user_access_key(self, username, access_key):
+        """
+        Deletes a user's access key
+        """
+        if self.dry_run:
+            return
+        self.iam_client.delete_access_key(
+            UserName=username,
+            AccessKeyId=access_key['AccessKeyId']
+        )
+
+    def delete_user(self, username):
+        """
+        Deletes a IAM user
+        """
+        if self.dry_run:
+            return
+        self.iam_client.delete_user(UserName=username)
+
+    def get_iam_users(self):
+        """
+        Lists all IAM users
+        """
+        users = []
+        paginator = self.iam_client.get_paginator('list_users')
+        for page in paginator.paginate():
+            for user in page['Users']:
+                users.append(user)
+        return users
+
+    def get_iam_user_old_access_keys(self, username):
+        """
+        Lists all IAM user access keys and returns only the ones that haven't
+        been used in at least age_limit
+        """
+        old_keys = []
+        user_access_keys = []
+
+        # Get all user access keys
+        paginator = self.iam_client.get_paginator('list_access_keys')
+        for page in paginator.paginate(UserName=username):
+            for access_key in page['AccessKeyMetadata']:
+                user_access_keys.append(access_key)
```

Comment:
@giovannicimolin, you could avoid using the `.extend()` method here and combine the checks in the next for loop into this one itself.

Reply:
You can also create a nested function which returns `True` or `False` based on whether a key is old or not and use a list comprehension in the for loop to construct the list. Check whichever way is more concise and pythonic and use it.

Reply:
@lgp171188 Something like this?
```
        old_keys = []

        def is_key_old(access_key):
            last_used = self.iam_client.get_access_key_last_used(AccessKeyId=access_key['AccessKeyId'])
            last_used_date = last_used.get('AccessKeyLastUsed', {}).get('LastUsedDate')
            if last_used_date and (last_used_date < self.age_limit):
                return True
            return False

        # Get all user access keys
        paginator = self.iam_client.get_paginator('list_access_keys')
        for page in paginator.paginate(UserName=username):
            old_keys2.extend([key for key in page['AccessKeyMetadata'] if is_key_old(key)])

        return old_keys
```